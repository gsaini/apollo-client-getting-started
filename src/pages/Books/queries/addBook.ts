import { gql } from "@apollo/client";

export const ADD_BOOK = gql`
  mutation addBook($title: String!, $author: String!) {
    addBook(title: $title, author: $author) {
      title
      author
    }
  }
`;
