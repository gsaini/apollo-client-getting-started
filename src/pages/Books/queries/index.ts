export * from "./addBook";
export * from "./getBooks";
