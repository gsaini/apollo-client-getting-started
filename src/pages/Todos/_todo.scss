.Todo {
  &__list {
    display: flex;
    flex-direction: column;
  }

  &__item {
    padding: 20px;
  }
}
