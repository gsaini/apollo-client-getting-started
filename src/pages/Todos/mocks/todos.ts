export const todos = [
  {
    id: "1",
    title: "delectus aut autem",
    completed: false,
    __typename: "ToDo",
  },
  {
    id: "2",
    title: "quis ut nam facilis et officia qui",
    completed: false,
    __typename: "ToDo",
  },
  {
    id: "3",
    title: "fugiat veniam minus",
    completed: false,
    __typename: "ToDo",
  },
  {
    id: "4",
    title: "et porro tempora",
    completed: true,
    __typename: "ToDo",
  },
  {
    id: "5",
    title: "laboriosam mollitia et enim quasi adipisci quia provident illum",
    completed: false,
    __typename: "ToDo",
  },
];
