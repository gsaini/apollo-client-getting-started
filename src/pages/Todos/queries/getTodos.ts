import { gql } from "@apollo/client";

export const GET_TODOS = gql`
  query GetTodos {
    todos {
      id
      title
      completed
    }
  }
`;
