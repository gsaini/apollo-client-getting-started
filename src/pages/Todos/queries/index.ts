export * from "./getTodos";
export * from "./updateTodo";
