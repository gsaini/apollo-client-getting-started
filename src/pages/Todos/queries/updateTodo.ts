import { gql } from "@apollo/client";

export const UPDATE_TODO = gql`
  mutation UpdateTodo($id: ID!, $completed: Boolean!) {
    updateTodo(id: $id, completed: $completed) {
      id
      completed
    }
  }
`;
