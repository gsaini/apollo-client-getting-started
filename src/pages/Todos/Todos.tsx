import React from 'react';
import { useQuery, useMutation } from '@apollo/client';
import { GET_TODOS, UPDATE_TODO } from "./queries";
import { TodoItem } from '../../components/TodoItem';
import { ToDo } from '../../types';
import './_todo.scss';

export const Todos = () => {
  const { loading, error, data } = useQuery(GET_TODOS);
  const [updateTodo] = useMutation(UPDATE_TODO);

  if (loading) return <p>Loading...</p>;
  if (error) return <p>Error :(</p>;

  /**
   * Toggle Todo item.
   * @param todo todo item
   */
  const onToggleTaskItem = ({ id, completed }): void => {
    updateTodo({ variables: { id, completed } });
  }

  return (
    <form className="Todo">
      <div className="Todo__list">
        {
          data.todos.map((todo: ToDo) => {
            return (
              <TodoItem key={`item-${todo.id}`} todo={todo} onChange={onToggleTaskItem} />
            );
          })
        }
      </div>
    </form>);
}