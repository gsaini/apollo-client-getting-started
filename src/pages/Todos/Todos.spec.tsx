import React from 'react';
import { render } from '../../test-utils';
import { generateImage } from 'jsdom-screenshot';
import { axe } from '../../test-utils';
import { Todos } from './Todos';
import { GET_TODOS, UPDATE_TODO } from './queries';
import { todos } from './mocks';
import { fireEvent, waitFor } from '@testing-library/react';

describe('Todos components', () => {
    const mocks = [{
        request: {
            query: GET_TODOS
        },
        result: {
            data: { todos }
        }
    }];

    it('should render snapshot', async () => {
        const { container } = await render(<Todos />, { mocks });
        expect(container).toMatchSnapshot();
    });

    it('should render visual snapshot', async () => {
        await render(<Todos />, { mocks });

        const screenshot = await generateImage();
        expect(screenshot).toMatchImageSnapshot();
    });

    it('should validate accessibility voilations', async () => {
        const { container } = await render(<Todos />, { mocks });
        expect(await axe(container)).toHaveNoViolations();
    });

    it('should update todo on toggle task selection', async () => {
        const mutationMock = [{
            request: {
                query: UPDATE_TODO,
                variables: { id: '1', completed: true }
            },
            result: jest.fn(() => ({
                data: {
                    updateTodo: {
                        ...todos[0],
                        completed: true
                    }
                }
            }))
        }];
        const { getByLabelText } = await render(<Todos />, { mocks: mutationMock });
        fireEvent.click(getByLabelText(todos[0].title));
        const updateMutationMock = mutationMock[0].result;
        await waitFor(() => expect(updateMutationMock).toHaveBeenCalledWith());
    });
});
