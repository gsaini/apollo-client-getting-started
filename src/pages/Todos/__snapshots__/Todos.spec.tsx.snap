// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos components should render snapshot 1`] = `
<div>
  <form
    class="Todo"
  >
    <div
      class="Todo__list"
    >
      <div
        class="Todo__item"
      >
        <label
          for="1-checkbox"
        >
          delectus aut autem
        </label>
        <input
          id="1-checkbox"
          type="checkbox"
        />
      </div>
      <div
        class="Todo__item"
      >
        <label
          for="2-checkbox"
        >
          quis ut nam facilis et officia qui
        </label>
        <input
          id="2-checkbox"
          type="checkbox"
        />
      </div>
      <div
        class="Todo__item"
      >
        <label
          for="3-checkbox"
        >
          fugiat veniam minus
        </label>
        <input
          id="3-checkbox"
          type="checkbox"
        />
      </div>
      <div
        class="Todo__item"
      >
        <label
          for="4-checkbox"
        >
          et porro tempora
        </label>
        <input
          checked=""
          id="4-checkbox"
          type="checkbox"
        />
      </div>
      <div
        class="Todo__item"
      >
        <label
          for="5-checkbox"
        >
          laboriosam mollitia et enim quasi adipisci quia provident illum
        </label>
        <input
          id="5-checkbox"
          type="checkbox"
        />
      </div>
    </div>
  </form>
</div>
`;
